cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -1107,17 +1107,9 @@ public void removeApplication(ApplicationReference appRef)
     // since already checked DELETE privileges above, no need to check again for appDetail
     // ensure no running programs across all versions of the application
     ensureNoRunningPrograms(appRef);
-
-    store.scanApplications(
-        ScanApplicationsRequest.builder().setApplicationReference(appRef).build(),
-        batchSize,
-        (appId, app) -> {
-          try {
-            deleteApp(appId, app.getSpec());
-          } catch (IOException e) {
-            throw new RuntimeException(e);
-          }
-        });
+    ApplicationId appId = new ApplicationId(appRef.getNamespace(), appRef.getApplication(),
+                                            appMeta.getSpec().getAppVersion());
+    deleteApp(appId, appMeta.getSpec());
```

Comment:
this method should take an ApplicationReference instead of an Id. Id should only be used for version specific operations.

Reply:
It looks like deleteApp() is still using a bunch of methods that take ApplicationId when it should take ApplicationReference. We should be changing all of these to use reference instead of id to make sure they are doing the right thing.

We can do this in a follow up PR, but please put a comment with a TODO here, otherwise the method signature of deleteApp() makes it seem like it's only deleting one version.